Projects/BookJSON2/main.go

```diff
@@ -20,96 +22,164 @@ type Author struct {
+
+					w.WriteHeader(http.StatusBadRequest)
+					fmt.Fprintf(w, "Book with Id=%v already exists in record", newbook.Id)
+					return
+				}
+			}
 			b.Books = append(b.Books, newbook)
-			json.NewEncoder(w).Encode(newbook)
-			break
+			fmt.Fprintln(w, "Added new Book")
 
+			if err := json.NewEncoder(w).Encode(newbook); err != nil {
+				log.Println(err)
+			}
 		}
-	case "DELETE": //Delete a specific book data
+
+	case http.MethodDelete: //Delete a specific book data
 		{
 			w.Header().Set("content-type", "application/json")
-			jsonrequestfromweb := Book{}
-			_ = json.NewDecoder(r.Body).Decode(&jsonrequestfromweb)
-			b.Books = append(b.Books[:jsonrequestfromweb.Id], b.Books[jsonrequestfromweb.Id+1:]...)
-			json.NewEncoder(w).Encode(b.Books)
-			break
+
+			id, err := strconv.Atoi(path[2])
+			if err != nil {
+				log.Println(err)
+				w.WriteHeader(http.StatusBadGateway)
+				fmt.Fprintln(w, "Specify a valid Id")
+				return
+			}
+
+			for index, match := range b.Books {
+				if match.Id == id {
+					b.Books = append(b.Books[:index], b.Books[index+1:]...)
+					fmt.Fprintln(w, "Deletion Successful")
+					return
+				}
+			}
+			w.WriteHeader(http.StatusNotFound)
+			fmt.Fprintf(w, "Book of Id %v doesn't exist\n", id)
+			return
 		}
-	case "PATCH": //Update an existing book
+	case http.MethodPut: //Update an existing book
 		{
+			if path[2] != "" {
+
+				w.WriteHeader(http.StatusBadRequest)
+				fmt.Fprintln(w, "Cannot specify any value while Updating")
+				return
+			}
 			w.Header().Set("content-type", "application/json")
 			jsonrequestfromweb := Book{}
 			_ = json.NewDecoder(r.Body).Decode(&jsonrequestfromweb)
-			b.Books = append(b.Books[:jsonrequestfromweb.Id], b.Books[jsonrequestfromweb.Id+1:]...)
-			b.Books = append(b.Books, jsonrequestfromweb)
-			json.NewEncoder(w).Encode(b.Books)
-			break
+
+			for index, match := range b.Books {
+				if match.Id == jsonrequestfromweb.Id {
+					b.Books = append(b.Books[:index], b.Books[index+1:]...)
+					b.Books = append(b.Books, jsonrequestfromweb)
+					fmt.Fprintln(w, "Update successful")
+					return
+
+				}
+
+			}
+			w.WriteHeader(http.StatusBadRequest)
+			fmt.Fprintln(w, "Please Enter Valid Book Id in json body")
+
+			return
 		}
 	}
 }
 
-func (b *Books) allbooks(w http.ResponseWriter, r *http.Request) {
+func (b *BookHandler) allbooks(w http.ResponseWriter, r *http.Request) {
 
 	switch r.Method {
 
-	case "GET": //fetch the list of all books
+	case http.MethodGet: //fetch the list of all books
 		{
-			w.Header().Set("content-type", "application/json")
-			json.NewEncoder(w).Encode(b.Books)
-			break
+
+			if err := json.NewEncoder(w).Encode(b.Books); err != nil {
+				log.Println(err)
+			}
+
 		}
-	case "DELETE": //delete all books
+	case http.MethodDelete: //delete all books
 		{
 			w.Header().Set("content-type", "application/json")
-			b.Books = nil
-			json.NewEncoder(w).Encode(b.Books)
-			break
 
+			b.Books = nil
+			fmt.Fprintln(w, "Library Cleared")
```

Comment:
you are posting above that the response is JSON (line 161) but this is not JSON.

Reply:
resolved with https://github.com/AmanSarraf/Go-Project/pull/9/commits/60e85cf7321c1abf9257e30d57a0ddccf479e689